### ISSUE TITLE
Introduce Continuous Integration

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#3115]
Key functionality:

* Builds triggered automatically on approved commits to the repo
* Each active branch made available on it's own URL
 * e.g. `https://tiddlywiki.com/*` is an alias of `https://tiddlywiki.com/branches/master/*`
 * The build products of these branches would be deleted when the branch is deleted
 * Prereleases would now be a branch called prerelease

Approach:

* All new scripts to be compatible with Bash and PowerShell
* All scripts to be accessed via npm

Tasks:

* [ ] @pmario will set up throwaway toy GitHub accounts to test the proposed setup
 * Done in that way it may also be a useful template for other people who need CI so maybe call it Testing2443242 or how-to-setup-continuous-integration
 * Anyone interested in helping should notify @pmario here so that he can add them to the new repos
 * Review and decide
* [ ] (to be planned after step 1) apply the same changes to TiddlyWiki5
 * [ ] @pmario et al A PR to make the necessary changes to the repo (and we need to decide what goes in build.jermolene.github.io vs. TiddlyWiki 5)
 * [ ] Clear step by step documentation of what needs to be done to set things up in Travis
 * [ ] @Jermolene to perform those steps in the tiddlywiki travis.org account

People:

@pmario @mklauber have expressed an interest in helping with this task, please add a comment if you'd like to be allocated some of the tasks too!

**Comment by ibnishak** [comment]
While I don't have the capacity to provide technical support, I would be willing offer my services in testing.

**Comment by pmario** [comment]
@ibnishak ... please have a look to video 1-4   or  up to 6/7 if you are interested.

**Comment by pmario** [comment]
uups missing link :) https://www.youtube.com/playlist?list=PLuiC_HFhI4OyUiDGqvzB64mTQTURABjlW

**Comment by Jermolene** [comment]
I've spent the last few days making some good progress on this: I've now got Travis CI building tiddlywiki.com/prerelease to a throwaway repository (https://github.com/Jermolene/Testing2019) each time a commit is pushed to the "master" branch.

The builds are the same apart from a couple of minor inadvertent differences:

* Some CSS ordering in the static renderings is different. I don't think it will have an impact
* The timezone for Travis CI builds is one hour different than my machine, and so the static renderings report a different time

There's also an improvement I've been wanting to make for a while: every release apart from empty.html includes a tiddler $:/build with content such as:

```
Prerelease built from branch 'master' at commit d8104e2c4197fe81b3b12588f13a60843a9359b3 of https://github.com/Jermolene/TiddlyWiki5.git at 2019-01-08 10:04:23 UTC
```

The TW5 changes involved can be seen here: https://github.com/Jermolene/TiddlyWiki5/compare/31470f6...master

This commit shows the differences between the prerelease built with the old method vs. the new CI method: https://github.com/Jermolene/Testing2019/commit/ab02a9067d4bc4a49ee40e54abf209b9d9412993

My next steps are to set up and test the non-prerelease build and then I think we're good to switch over.

I'd appreciate any assistance with improving my `.travis.yml`: the worst part of doing this work was endless frustrating Googling to find out how to do things in both Travis CI and YAML. Once we've deployed the current stuff I'd also like to work on making the builds faster -- it's currently about 6 mins.

**Comment by BurningTreeC** [comment]
Hi @Jermolene, big work :smile: !

I have a question regarding "moving over", does that include moving to the TiddlyWiki repo?

I'm asking because you mentioned in the google groups, answering in my "Simple Android App" thread, that we could create a repo for it and I'm at the point where I'd like to do so, but putting only the basic skeletal structure on it and building up on that with (hopefully) some other interested people.

Should I create a separate issue somewhere?

**Comment by Jermolene** [comment]
Hi @BurningTreeC 

> I have a question regarding "moving over", does that include moving to the TiddlyWiki repo?

I decided to simplify things by doing them one at a time. I still plan to move TW5 to the TiddlyWiki org, but plan to do it with TiddlyDesktop first to see how it goes.

> I'm asking because you mentioned in the google groups, answering in my "Simple Android App" thread, that we could create a repo for it and I'm at the point where I'd like to do so, but putting only the basic skeletal structure on it and building up on that with (hopefully) some other interested people.

That's great; there's nothing to stop us making a repo in the TiddlyWiki repo for the SAA. Does it not have a name yet? TiddlyMobile or TiddlyDroid might be a good match for TiddlyDesktop, perhaps.

**Comment by BurningTreeC** [comment]
>I decided to simplify things by doing them one at a time. I still plan to move TW5 to the TiddlyWiki org, but plan to do it with TiddlyDesktop first to see how it goes.

That's cool for me :+1: 

>That's great; there's nothing to stop us making a repo in the TiddlyWiki repo for the SAA. Does it not have a name yet? TiddlyMobile or TiddlyDroid might be a good match for TiddlyDesktop, perhaps.

No I'm not settled with a name. I think the final app should show up with a short, catchy name below its icon and the app should keep that naming. TiddlyDroid or TiddlyMobile would be good for repo-names - if we can go with a different naming for the app internally... but I don't want to block this just because of being unable to give it a name. I called it "TW" and I like how that appears on my phone

**Comment by Jermolene** [comment]
How about AndTW? It's short, and nods to the previous app.

**Comment by BurningTreeC** [comment]
How about aTW?

**Comment by Jermolene** [comment]
> How about aTW?

Perfect!

**Comment by BurningTreeC** [comment]
>How about aTW?

just because I don't like those And* *Droid and so on

Ok let's go with that, thanks Jeremy!

**Comment by Jermolene** [comment]
> Ok let's go with that, thanks Jeremy!

Shall I go ahead and create the repo and add you to it?

**Comment by BurningTreeC** [comment]
>Shall I go ahead and create the repo and add you to it?

Yes please, I don't know much about the next steps to take but I'll try figuring it out.
You'll need to be a bit lenient with me in the beginnings

**Comment by BurningTreeC** [comment]
... or "TWApp" which is the name my current (testing-)repo uses

**Comment by BurningTreeC** [comment]
>Hi @BurningTreeC I created the repo as "aTW" because I think "TWApp" isn't distinctive enough; I commonly use it as a generic throwaway app name when trying out TiddlyWiki stuff.

That's great, thank you @Jermolene !

I hope as many people as possible come over helping to make it become something

Thanks for the hint about the license, I'll copy that over first

**Comment by BurningTreeC** [comment]
again, @Jermolene - in the license, what do you suggest for:

- line 1: `TiddlyWiki created by Jeremy Ruston, (jeremy [at] jermolene [dot] com)`
- following lines, the dates and copyrights:

```
Copyright (c) 2004-2007, Jeremy Ruston
Copyright (c) 2007-2018, UnaMesa Association
All rights reserved.
```

thanks for your patience

**Comment by Jermolene** [comment]
I'd suggest that you keep the line 1 about TiddlyWiki, adding the word "Incorporating", and precede it with a new line saying `aTW created by $YOU, ($OBFUSCATED EMAIL)`. And then the copyright notice should just be UnaMesa if you're happy to donate it (feel free to keep your own copyright, too; the motivation for me passing the TW copyrights to UnaMesa was that ownership by a non-profit presents fewer hurdles for people thinking of adopting the software).
